drivers/blk/sdmmc/sdmmc_protocol/sdmmc.rs

```diff
@@ -0,0 +1,405 @@
+    Success = 0b0001,    // 1st bit
+    Error   = 0b0010,    // 2nd bit
+    SDIO    = 0b0100,    // 3rd bit
+    // You can add more flags as needed
+}
+
+// Define the MMC response flags
+const MMC_RSP_PRESENT: u32 = 1 << 0;
+const MMC_RSP_136: u32 = 1 << 1;       // 136-bit response
+const MMC_RSP_CRC: u32 = 1 << 2;       // Expect valid CRC
+const MMC_RSP_BUSY: u32 = 1 << 3;      // Card may send busy
+const MMC_RSP_OPCODE: u32 = 1 << 4;    // Response contains opcode
+
+// Define the MMC response types
+pub const MMC_RSP_NONE: u32 = 0;
+pub const MMC_RSP_R1: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R1B: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE | MMC_RSP_BUSY;
+pub const MMC_RSP_R2: u32 = MMC_RSP_PRESENT | MMC_RSP_136 | MMC_RSP_CRC;
+pub const MMC_RSP_R3: u32 = MMC_RSP_PRESENT;
+pub const MMC_RSP_R4: u32 = MMC_RSP_PRESENT;
+pub const MMC_RSP_R5: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R6: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+pub const MMC_RSP_R7: u32 = MMC_RSP_PRESENT | MMC_RSP_CRC | MMC_RSP_OPCODE;
+
+
+/// Program async Rust can be very dangerous if you do not know what is happening understand the hood
+/// Power up and power off cannot be properly implemented if I do not have access to control gpio/ regulator and timer
+pub trait SdmmcHardware {
+    fn sdmmc_power_up(&mut self) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_set_ios(&mut self) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_send_command(&mut self, cmd: &SdmmcCmd, data: Option<&MmcData>) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_receive_response(&self, cmd: &SdmmcCmd, response: &mut [u32; 4]) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_enable_interrupt(&mut self, irq_to_enable: &mut u32) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_ack_interrupt(&mut self, irq_enabled: &u32) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+
+    fn sdmmc_power_off(&mut self) -> Result<(), SdmmcHalError> {
+        return Err(SdmmcHalError::ENOTIMPLEMENTED);
+    }
+}
+
+// Not used right now, but would be useful in the future once we want to execute some command synchronously
+fn send_cmd_and_receive_resp<T: SdmmcHardware>(
+    hardware: &mut T,
+    cmd: &SdmmcCmd,
+    data: Option<&MmcData>,
+    resp: &mut [u32; 4]
+) -> Result<(), SdmmcHalError> {
+    // Send the command using the hardware layer
+    let mut res = hardware.sdmmc_send_command(cmd, data);
+    if res.is_err() {
+        return res;
+    }
+
+    // TODO: Change it to use the sleep function provided by the hardware layer
+    let mut retry: u32 = 1000000;
```

Comment:
This is a lot of retries... Is there a particular reason you chose this number?

Reply:
This is a busy poll retry, we could poll infinitely if we trust the device to be correct. I will add a comment to explain this.

Reply:
>Not clear if this is something that we actually want or not...

From the designer's perspective, not all the functions defined in traits will need to be implemented for the driver to work. There will be a function to ask for the hardware layer to provide all of its abilities and the protocol layer to determine if those satisfy the minimal requirement to get the driver working and if so, protocol layer try to achieve the maximum speed possible based on the capabilities provided. I would hope to have more clarification on this one.